cmake/WarningsUtils.cmake

```diff
@@ -0,0 +1,49 @@
+#===============================================================================
+# Copyright 2023 Codeplay Software Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+#
+#
+# SPDX-License-Identifier: Apache-2.0
+#===============================================================================
+
+include_guard(GLOBAL)
+
+add_library(onemkl_warnings INTERFACE)
+
+set(ONEMKL_WARNINGS "")
+
+include(CheckCXXCompilerFlag)
+macro(add_warning flag)
+  check_cxx_compiler_flag(${flag} IS_SUPPORTED)
+  message(STATUS "DBG '${flag}': ${IS_SUPPORTED}")
+  if(${IS_SUPPORTED})
+    list(APPEND ONEMKL_WARNINGS ${flag})
+  else()
+    message(WARNING "Compiler does not support ${flag}")
+  endif()
+endmacro()
+
+add_warning("-Wall")
+add_warning("-Wextra")
+add_warning("-Wshadow")
+add_warning("-Wconversion")
+add_warning("-Wpedantic")
+
+message(STATUS "Using warnings: ${ONEMKL_WARNINGS}")
+
+# The onemkl_warnings target can be linked to any other target to enable warnings.
+target_compile_options(onemkl_warnings INTERFACE ${ONEMKL_WARNINGS})
+
+# Add the library to install package
+install(TARGETS onemkl_warnings EXPORT oneMKLTargets)
```

Comment:
Why does this need to be installed?

Reply:
CMake is failing otherwise:
```
CMake Error: install(EXPORT "oneMKLTargets" ...) includes target "onemkl_dft_mklgpu_obj" which requires target "onemkl_warnings" that is not in any export set.
CMake Error: install(EXPORT "oneMKLTargets" ...) includes target "onemkl_dft" which requires target "onemkl_warnings" that is not in any export set.
CMake Error in CMakeLists.txt:
  export called with target "onemkl_dft_mklgpu_obj" which requires target
  "onemkl_warnings" that is not in any export set.


CMake Error in CMakeLists.txt:
  export called with target "onemkl_dft" which requires target
  "onemkl_warnings" that is not in any export set.
```

I'm not entirely sure why. I've made sure to link this target with `PRIVATE` so it does not affect users linking with oneMKL. Maybe CMake is not smart enough to detect that `onemkl_warnings` is not really needed?